x509-cert/src/certificate.rs

```diff
@@ -143,31 +143,85 @@ pub struct TbsCertificateInner<P: Profile + 'static = Rfc5280> {
     /// require later versions. Care should be taken in order to ensure
     /// standards compliance.
     #[asn1(context_specific = "0", default = "Default::default")]
-    pub version: Version,
+    pub(crate) version: Version,
 
-    pub serial_number: SerialNumber<P>,
-    pub signature: AlgorithmIdentifierOwned,
-    pub issuer: Name,
-    pub validity: Validity<P>,
-    pub subject: Name,
-    pub subject_public_key_info: SubjectPublicKeyInfoOwned,
+    pub(crate) serial_number: SerialNumber<P>,
+    pub(crate) signature: AlgorithmIdentifierOwned,
+    pub(crate) issuer: Name,
+    pub(crate) validity: Validity<P>,
+    pub(crate) subject: Name,
+    pub(crate) subject_public_key_info: SubjectPublicKeyInfoOwned,
 
     #[asn1(context_specific = "1", tag_mode = "IMPLICIT", optional = "true")]
-    pub issuer_unique_id: Option<BitString>,
+    pub(crate) issuer_unique_id: Option<BitString>,
 
     #[asn1(context_specific = "2", tag_mode = "IMPLICIT", optional = "true")]
-    pub subject_unique_id: Option<BitString>,
+    pub(crate) subject_unique_id: Option<BitString>,
 
     #[asn1(context_specific = "3", tag_mode = "EXPLICIT", optional = "true")]
-    pub extensions: Option<crate::ext::Extensions>,
+    pub(crate) extensions: Option<ext::Extensions>,
 }
 
 impl<P: Profile> TbsCertificateInner<P> {
+    /// Version of this certificate.
+    pub fn version(&self) -> Version {
+        self.version
+    }
+
+    /// Serial number of this certificate.
+    pub fn serial_number(&self) -> &SerialNumber<P> {
+        &self.serial_number
+    }
+
+    /// Signature algorithm identifier.
+    pub fn signature(&self) -> &AlgorithmIdentifierOwned {
+        &self.signature
+    }
+
+    /// Certificate issuer.
+    pub fn issuer(&self) -> &Name {
+        &self.issuer
+    }
+
+    /// Validity period for this certificate.
+    pub fn validity(&self) -> &Validity<P> {
+        &self.validity
+    }
+
+    /// Subject of this certificate.
+    pub fn subject(&self) -> &Name {
+        &self.subject
+    }
+
+    /// Subject Public Key Info (SPKI): public key information about this certificate including
+    /// algorithm identifier and key data.
+    pub fn subject_public_key_info(&self) -> &SubjectPublicKeyInfoOwned {
+        &self.subject_public_key_info
+    }
+
+    /// Issuer unique ID.
+    pub fn issuer_unique_id(&self) -> &Option<BitString> {
+        &self.issuer_unique_id
+    }
+
+    /// Subject unique ID.
+    pub fn subject_unique_id(&self) -> &Option<BitString> {
+        &self.subject_unique_id
+    }
+
+    /// Certificate extensions.
+    pub fn extensions(&self) -> &ext::Extensions {
```

Comment:
I'm curious about the move to non-optional list of extensions here.
I think I would see the value of keeping the option.

Reply:
There were several test examples which moved the extensions out of the struct, which isn't possible with an immutable borrow.

I first went down the path of cloning/unwrapping, but the ergonomics for accessing the extensions are a lot simpler this way.

What is use case for `Option::None` versus an empty set, particularly within the context of a public accessor method? Is there something semantically meaningful to distinguish there?

Reply:
I don't know. There is a semantic in an `Option<Vec<Extension>>` as this represents how this is serialized.

`None` will skip the field altogether, `Some<empty vec>` will put an empty sequence.

Is there value for a consumer? I'm not entirely sure.

Reply:
In the interest of moving this along and keeping the changes minimal I changed it to return `Option<&ext::Extension>` in [0f9b52b](https://github.com/RustCrypto/formats/pull/1505/commits/0f9b52bc3f13a8eb1dd6fb92187d2745685684a8), however I think it would still be good to circle back on this in a separate PR